### PR TITLE
Relationship Attribute Usage, Relationship Attribute Name, "do" helper script

### PR DESCRIPTION
**Pull request by dave-cheong**

### CODE REVIEW

File: .node-version [hunk 1]
```diff
@@ -0,0 +1 @@
+6.1.0
```
Comment (sundriver) [inline .node-version]: Do we need this file ?

File: backend/gulpfile.js [hunk 2]
```diff
@@ -85,14 +86,20 @@ gulp.task("publish:tarball",
...
-        jasmine({
-            verbose: true,
-            includeStackTrace: true,
-            config: {
-                stopSpecOnExpectationFailure: false,
-                random: false
-            }
-        })
-    );
+  var pattern = ['dist/{**,./}/*.spec.js'];
```
Comment (sundriver) [inline backend/gulpfile.js]: Adding comments as how to use this
--test=sometest => execute sometest.spec.js
Good place to document it is https://github.com/atogov/RAM/blob/develop/backend/readme.md

File: backend/typescript/models/base.ts [hunk 3]
```diff
@@ -16,6 +18,7 @@ export interface IRAMObject extends mongoose.Document {
   delete(): void;
 }
 
+/* tslint:disable:max-func-body-length */
```
Comment (sundriver) [inline backend/typescript/models/base.ts]: I am not sure if you need this in this case, too many empty lines

File: backend/typescript/models/relationshipAttributeName.model.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    RelationshipAttributeNameBooleanDomain,
+    RelationshipAttributeNameNumberDomain,
+    RelationshipAttributeNameStringDomain,
+    RelationshipAttributeNameDateDomain,
+    RelationshipAttributeNameSingleSelectDomain,
+    RelationshipAttributeNameMultiSelectDomain
+];
+
+export interface IRelationshipAttributeName extends ICodeDecode {
+    domain: string;
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeName.model.ts]: Adding comment what domain presents in the code

File: backend/typescript/models/relationshipAttributeName.model.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    purposeText: string;
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeName.model.ts]: I can't find this in data model

File: backend/typescript/models/relationshipAttributeNameUsage.model.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export interface IRelationshipAttributeNameUsage extends mongoose.Document {
+    optionalInd: boolean;
+    defaultValue: string;
+    attributeName: IRelationshipAttributeName;
+}
+
+const RelationshipAttributeNameUsageSchema = new mongoose.Schema({
+    optionalInd: {
+        type: Boolean,
+        default: false
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeNameUsage.model.ts]: Anything that has default should be 
required: true 
otherwise null is accepted as a valid value, please apply to all the code

File: .node-version (hunk 1 above)
Comment (dave-cheong) [inline .node-version]: Depends on whether you are using a node version manager or not, like avn. We will be standardising on the node version, so let's chat about the version manager to use.

File: backend/gulpfile.js (hunk 2 above)
Comment (dave-cheong) [inline backend/gulpfile.js]: This is documented in the ./do script

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 4 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeName.model.ts]: https://github.com/atogov/RAM/wiki/Relationship-Attribute-Types

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 5 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeName.model.ts]: https://github.com/atogov/RAM/wiki/Relationship-Attribute-Types

File: backend/typescript/models/relationshipAttributeNameUsage.model.ts (hunk 6 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeNameUsage.model.ts]: Ok, updated

File: backend/gulpfile.js (hunk 2 above)
Comment (sundriver) [inline backend/gulpfile.js]: Can we have it documented on the website as well,
Nobody knows about it ATM
That will promote people to use it

File: tests/tsconfig.json [hunk 7]
```diff
@@ -15,6 +15,9 @@
         "typings",
         "node_modules"
     ],
+    "files": [
+        "./Boot.ts"
```
Comment (sundriver) [inline tests/tsconfig.json]: When I run tsc in test folder I get:
error TS6053: File 'Boot.ts' not found.

File: tests/typescript/Boot.ts [hunk 8]
```diff
@@ -0,0 +1,2 @@
+/// <reference path="../typings/index.d.ts" />
```
Comment (sundriver) [inline tests/typescript/Boot.ts]: Why we need this file, it is empty !

File: tests/typescript/steps/relationshipTypes.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import {RestCalls} from '../support/rest';
+import {config} from '../bootstrap';
+
+const rest = new RestCalls(config.host, config.port);
+
+export default class RelationshipTypeSteps {
```
Comment (sundriver) [inline tests/typescript/steps/relationshipTypes.ts]: I can't find this in the data model.
Add jsdoc documentation please

File: backend/gulpfile.js (hunk 2 above)
Comment (dave-cheong) [inline backend/gulpfile.js]: Updated README with the "The do script" section.

File: backend/typescript/models/base.ts (hunk 3 above)
Comment (dave-cheong) [inline backend/typescript/models/base.ts]: This might grow over time and in this case this function is now doing quite a few different things, which I'm happy to keep together. So let's use line breaks as a means to get some clarity.

File: tests/typescript/Boot.ts (hunk 8 above)
Comment (paulrule) [inline tests/typescript/Boot.ts]: I've moved the reference to bootstrap.ts (required so typescript can find the jamsine types) so I could delete the Boot.ts.

File: tests/tsconfig.json (hunk 7 above)
Comment (paulrule) [inline tests/tsconfig.json]: I've deleted Boot.ts and removed the reference from tsconfig.json

File: tests/typescript/steps/relationshipTypes.ts (hunk 9 above)
Comment (paulrule) [inline tests/typescript/steps/relationshipTypes.ts]: steps/relationshipTypes.ts are test 'Steps' for relationshipType - they provide access to the REST API for reuse by any tests that might require it.

I've renamed it to /steps/relationshipTypeSteps.ts to avoid confusion and added comments:

/**
 * Reusable test steps for the RelationshipType REST API.
 * 
 * Use these steps for any tests that need to interact with the RelationshipType REST API.
 */

File: do [hunk 10]
```diff
@@ -0,0 +1,218 @@
+#!/bin/bash
```
Comment (sundriver) [inline do]: Can we rename this file to something like ram ?
what does 'do' mean ?

File: .node-version (hunk 1 above)
Comment (sundriver) [inline .node-version]: which programs uses/applies this ?
If there is a program/script which is part of uses this, I am fine with it, need to documented

File: backend/gulpfile.js (hunk 2 above)
Comment (sundriver) [inline backend/gulpfile.js]: this special use case of test is not in the documentation

File: backend/typescript/models/base.ts (hunk 3 above)
Comment (sundriver) [inline backend/typescript/models/base.ts]: Let added it we it grows

File: do (hunk 10 above)
Comment (dave-cheong) [inline do]: Pushed.

File: backend/typescript/models/base.ts (hunk 3 above)
Comment (dave-cheong) [inline backend/typescript/models/base.ts]: We should monitor this, I'm not a big fan of keep docs separate from the doc. It means we have to keep them in sync and the script itself is already self documenting.

File: backend/gulpfile.js (hunk 2 above)
Comment (dave-cheong) [inline backend/gulpfile.js]: It is now

File: .node-version (hunk 1 above)
Comment (dave-cheong) [inline .node-version]: It is done automatically via avn.

Comment (sundriver) [inline .node-version]: we are not using avn.
I am happy if it was used somewhere in the ram script or somewhere people can use it.

File: backend/gulpfile.js (hunk 2 above)
Comment (sundriver) [inline backend/gulpfile.js]: Thanks

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 5 above)
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeName.model.ts]: it is not in data model

File: backend/typescript/models/relationshipAttributeNameUsage.model.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
+import * as mongoose from 'mongoose';
+import {IRelationshipAttributeName, RelationshipAttributeNameModel} from './relationshipAttributeName.model';
+
+// force schema to load first
+/* tslint:disable:no-unused-variable */
+const _RelationshipAttributeNameModel = RelationshipAttributeNameModel;
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeNameUsage.model.ts]: All tests pass without this line, 
could you explain why it is needed ? 
It doesn't generate anything in the final .js file !

File: backend/typescript/models/relationshipType.model.ts [hunk 12]
```diff
@@ -1,36 +1,54 @@
 import * as mongoose from 'mongoose';
 import {ICodeDecode, CodeDecodeSchema} from './base';
+import {IRelationshipAttributeNameUsage, RelationshipAttributeNameUsageModel} from './relationshipAttributeNameUsage.model';
 
-/* tslint:disable:no-var-requires */
-const mongooseIdValidator = require('mongoose-id-validator');
+// force schema to load first
+/* tslint:disable:no-unused-variable */
+const _RelationshipAttributeNameUsageModel = RelationshipAttributeNameUsageModel;
```
Comment (sundriver) [inline backend/typescript/models/relationshipType.model.ts]: same as above

File: backend/typescript/seeding/seed.ts [hunk 13]
```diff
@@ -87,5 +225,5 @@ const loadRelationshipTypes = async () => {
 Seeder
     .connect()
     .then(Seeder.dropDatabase)
-    .then(loadRelationshipTypes)
+    .then(load)
```
Comment (sundriver) [inline backend/typescript/seeding/seed.ts]: Please clearly name the function, load?

Comment (dave-cheong) [inline backend/typescript/seeding/seed.ts]: Renamed

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import * as mongoose from 'mongoose';
+import {ICodeDecode, CodeDecodeSchema} from './base';
+
+export interface IRelationshipAttributeName extends ICodeDecode {
+    domain: string;
+    purposeText: string;
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: I can't purposeText in data model

File: backend/typescript/models/relationshipType.model.ts (hunk 12 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipType.model.ts]: You need this or it will fail the test. You basically need to source the nested mongoose schema first before referencing it and the import alone doesn't do anything. You need to include it into the code first.

For example running

_./ram test:backend --test relationshipType.model_

Results in this error:

`MissingSchemaError: Schema hasn't been registered for model "RelationshipAttributeName".
`

File: backend/typescript/models/relationshipAttributeNameUsage.model.ts (hunk 11 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeNameUsage.model.ts]: Same as below.

File: backend/typescript/models/relationshipType.model.ts (hunk 12 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipType.model.ts]: So in the case of just running all tests via

`./ram test:backend`

It works, because you see the `RAM Relationship Attribute Name` tests running BEFORE `RAM Relationship Type` which is just a coincidence. So the const declaration to bring it in via the import is required.

File: backend/typescript/tests/relationshipType.model.spec.ts [hunk 15]
```diff
@@ -127,27 +143,27 @@ describe('RAM Relationship Type', () => {
...
         } catch (e) {
-            fail(e);
+            fail('Because ' + e);
+            done();
         }
-
     });
 
-    it('fails inserts with empty code', async (done) => {
+    it('fails insert with empty code', async (done) => {
```
Comment (sundriver) [inline backend/typescript/tests/relationshipType.model.spec.ts]: Actually this is testing it against null
We need to have minimum size for code in ICodeDecode

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    }
+});
+
+/* tslint:disable:no-empty-interfaces */
+export interface IRelationshipAttributeNameModel extends mongoose.Model<IRelationshipAttributeName> {
+    findByCode: (id:String) => mongoose.Promise<IRelationshipAttributeName>;
+    findValidByCode: (id:String) => mongoose.Promise<IRelationshipAttributeName>;
+}
+
+RelationshipAttributeNameSchema.static('findByCode', (code:String) => {
```
Comment (paulmarrington) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: I had trouble understanding the difference and I can picture times when the wrong one is called. Perhaps something like

findByCodeIgnoringDateRange and;
findByCodeInDateRange

Since the second is the myst likely to be used, using ignoring in the first highlights it's unusual usage.

File: tests/gulpfile.js [hunk 17]
```diff
@@ -2,41 +2,52 @@ var gulp = require('gulp');
 var jasmine = require('gulp-jasmine');
 var tslint = require('gulp-tslint');
 var ts = require('gulp-typescript');
+var sourcemaps = require("gulp-sourcemaps");
```
Comment (sundriver) [inline tests/gulpfile.js]: Doesn't follow agreed formating

File: tests/gulpfile.js [hunk 18]
```diff
@@ -2,41 +2,52 @@ var gulp = require('gulp');
...
 
-gulp.task('ts:compile',['ts:lint'], function () {
-    return gulp.src(['typescript/{**,./}/*.ts'])
-        .pipe(ts(tsProject, { sortOutput: true }))
-        .pipe(gulp.dest('dist/'));
+gulp.task("ts:compile", ["ts:lint"], function () {
+  var tsResult = gulp.src([
+    "typescript/**/*.ts"
+  ])
+    .pipe(sourcemaps.init())
```
Comment (sundriver) [inline tests/gulpfile.js]: why do we need sourcemaps in test?

File: tests/typescript/steps/relationshipTypeSteps.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+import {config} from '../bootstrap';
+
+const rest = new RestCalls(config.host, config.port);
+
+/**
+ * Reusable test steps for the RelationshipType REST API.
+ *
+ * Use these steps for any tests that need to interact with the RelationshipType REST API.
+ */
+export default class RelationshipTypeSteps {
```
Comment (sundriver) [inline tests/typescript/steps/relationshipTypeSteps.ts]: based on the comment, shouldn't this be called RelationshipTypeHelpers

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 5 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeName.model.ts]: Enforced in the code, per Dave Loone

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts (hunk 14 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: It's there - https://raw.githubusercontent.com/atogov/RAM/develop/docs/images/data-model.png

File: .node-version (hunk 1 above)
Comment (paulrule) [inline .node-version]: I've removed it and will just use it as a locally git ignored file. No time to debate this or work on other scripts right now.

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts (hunk 16 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: Let's have a separate discussion about this later. Disregard for this PR.

File: tests/typescript/steps/relationshipTypeSteps.ts (hunk 19 above)
Comment (paulrule) [inline tests/typescript/steps/relationshipTypeSteps.ts]: Okay, I'll rename them. You happy with /helpers/RelationshipTypeHelper.ts ?

File: backend/typescript/tests/relationshipType.model.spec.ts (hunk 15 above)
Comment (dave-cheong) [inline backend/typescript/tests/relationshipType.model.spec.ts]: Added more tests for null vs empty

File: backend/typescript/models/relationshipType.model.ts (hunk 12 above)
Comment (sundriver) [inline backend/typescript/models/relationshipType.model.ts]: Ok, please include this discussion in the comments of the code so we also follow it and don't miss it

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts (hunk 16 above)
Comment (sundriver) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: Dave, we don't understand if the naming is not descriptive, think about 6 months from now someone reviewing this code ...

File: backend/typescript/tests/relationshipType.model.spec.ts (hunk 15 above)
Comment (sundriver) [inline backend/typescript/tests/relationshipType.model.spec.ts]: Thanks

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 4 above)
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeName.model.ts]: Could you document this class
Also provide reference to wiki

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 5 above)
Comment (sundriver) [inline backend/typescript/models/relationshipAttributeName.model.ts]: ok

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts (hunk 14 above)
Comment (sundriver) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: Cool, thanks

File: tests/typescript/steps/relationshipTypeSteps.ts (hunk 19 above)
Comment (sundriver) [inline tests/typescript/steps/relationshipTypeSteps.ts]: Sure, go ahead

Comment (dave-cheong) [inline tests/typescript/steps/relationshipTypeSteps.ts]: Already done.

File: backend/typescript/models/relationshipAttributePermittedValue.model.ts (hunk 16 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributePermittedValue.model.ts]: Ok changed.

File: backend/typescript/models/relationshipType.model.ts (hunk 12 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipType.model.ts]: Done

File: backend/typescript/models/relationshipAttributeName.model.ts (hunk 4 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttributeName.model.ts]: Done
